ext/vl53/vl53_transport.hpp

```diff
@@ -0,0 +1,140 @@
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+// ----------------------------------------------------------------------------
+
+#ifndef MODM_VL53_TRANSPORT_HPP
+#define MODM_VL53_TRANSPORT_HPP
+
+#include <array>
+#include <cstdint>
+#include <modm/architecture/interface/delay.hpp>
+#include <modm/architecture/interface/i2c_device.hpp>
+#include <modm/architecture/interface/spi_device.hpp>
+#include <modm/processing/fiber.hpp>
+#include <optional>
+#include <span>
+
+namespace modm
+{
+
+class Vl53TransportBase
+{
+public:
+	virtual ~Vl53TransportBase() = default;
+
+	virtual uint8_t
+	writeMulti(uint16_t reg, const uint8_t *values, uint32_t size) = 0;
+	virtual uint8_t
+	readMulti(uint16_t reg, uint8_t *values, uint32_t size) = 0;
+	virtual uint8_t
+	writeByte(uint16_t reg, uint8_t value) = 0;
+	virtual uint8_t
+	readByte(uint16_t reg, uint8_t *value) = 0;
+	virtual uint8_t
+	resetSensor() = 0;
+};
+
+template<typename SpiMaster, typename Cs, typename Lpn, size_t VL53LMZ_COMMS_CHUNK_SIZE = 4096>
+class Vl53SpiTransport : public modm::SpiDevice<SpiMaster>, public Vl53TransportBase
+{
+public:
+	Vl53SpiTransport()
+	{
+		this->attachConfigurationHandler([] {
+			SpiMaster::setDataMode(SpiMaster::DataMode::Mode3);
+			SpiMaster::setDataOrder(SpiMaster::DataOrder::MsbFirst);
+			SpiMaster::setDataSize(SpiMaster::DataSize::Bit8);
+		});
+	};
+
+	Vl53SpiTransport(const Vl53SpiTransport &) = delete;
+
+	Vl53SpiTransport &
+	operator=(const Vl53SpiTransport &) = delete;
+
+	uint8_t
+	writeMulti(uint16_t register_address, const uint8_t *p_values, uint32_t size) override;
+
+	uint8_t
+	readMulti(uint16_t register_address, uint8_t *p_values, uint32_t size) override;
+
+	uint8_t
+	readByte(uint16_t register_address, uint8_t *p_value) override
+	{
+		return this->readMulti(register_address, p_value, 1);
+	}
+
+	uint8_t
+	writeByte(uint16_t register_address, uint8_t value) override
+	{
+		return this->writeMulti(register_address, &value, 1);
+	}
+
+	uint8_t
+	resetSensor() override;
+
+private:
+	uint8_t tx_buffer[VL53LMZ_COMMS_CHUNK_SIZE + 2];
+	uint8_t rx_buffer[VL53LMZ_COMMS_CHUNK_SIZE + 2];
+
+	static uint16_t
+	SPI_WRITE_MASK(uint16_t x) noexcept
+	{
+		return static_cast<uint16_t>(x | 0x8000);
+	}
+
+	static uint16_t
+	SPI_READ_MASK(uint16_t x) noexcept
+	{
+		return static_cast<uint16_t>(x & static_cast<uint16_t>(~0x8000));
+	}
+};
+
+template<typename I2cMaster, typename Lpn>
+class Vl53I2cTransport : public modm::I2cDevice<I2cMaster>, public Vl53TransportBase
+{
+public:
+	Vl53I2cTransport(const uint8_t address) : I2cDevice<I2cMaster>{address}, address_{address} {};
+
+	Vl53I2cTransport(const Vl53I2cTransport &) = delete;
+
+	Vl53I2cTransport &
+	operator=(const Vl53I2cTransport &) = delete;
+
+	inline void
+	SetAddress(const uint8_t address)
+	{
+		this->address_ = address;
```

Comment:
Isn't this functionality already [part of `modm::I2cDevice`](https://github.com/modm-io/modm/blob/develop/src/modm/architecture/interface/i2c_device.hpp#L58)?